bbbc021_feature_reduction_tests.py

```diff
@@ -0,0 +1,247 @@
+
+    ## run illumination correction on images
+   # for i, directory_files in enumerate(filenames):
+   #     run_illum(directory_files, names_illum[i])
+   #     print("Directory processed: ", directory_files[0])
+   
+    all_names_montage = []
+    ## montage illumed images
+    for i, filename_dir in enumerate(names_illum): 
+         sorted_filenames = sorted(filename_dir)
+         names_montage = make_montage_names(sorted_filenames)
+         all_names_montage.extend(names_montage)
+   #      run_montage(sorted_filenames, names_montage)
+   #      print("Directory {} out of {} processed".format(i, len(names_illum)))
+
+
+     ## run features on images
+    ims = map(io.imread, all_names_montage)
+    output_features(ims, all_names_montage, FEATURES_FILE)
+    
+     ## get x y coordinates
+    coords_pca = pca_transform(FEATURES_FILE)
+    print("Coords PCA retrieved")
+     ## generate CSV of coordinates
+    generate_bokeh_csv(coords_pca, all_names_montage, DATA_FILE)
+
+    coords_tsne = tsne_transform(FEATURES_FILE)
+    print("Coords tsne retrieved")
+    generate_bokeh_csv(coords_tsne, all_names_montage, DATA_TSNE)
+
+    coords_umap = umap_transform(FEATURES_FILE)
+    print("Coords umap retrieved")
+    generate_bokeh_csv(coords_umap, all_names_montage, DATA_UMAP)
+
+def make_montage_names(filenames):
+    """
+    Construct filenames for the montaged images
+
+    :param filenames: names of the illumed images with separate quadrants and channels
+    :return names_montage: names for each file which will result from montaging
+    """
+    filename_reg = r'(.*Week._.*_)(...)(_s._w..*)(_illum)(\.tif)$'
+    names_montage = []
+    for filename in filenames:
+        match = re.search(filename_reg, filename)
+        if match:
+            potential_name = match.group(1) + match.group(2) + "_montaged" + match.group(5)
+            if potential_name not in names_montage:
+                names_montage.append(potential_name)
+    return names_montage
+
+def get_valid_file_names(filepath):
+    """
+    Get full filenames relative to top level directory for each file in the BBBC trial, and
+    construct filenames with paths for saving output
+
+    :param filepath: path to the directory containing folders of images
+    :return (valid_filenames, illum_filenames): tuple with lists of filenames for reading and saving
+                                                                    images
+    """
+    filename_reg = r'(^Week._.*)(_..._s._w.)(.*)(\.tif)$'
+
+    valid_filenames = []
+    illum_filenames = []
+
+    for root, directories, filenames in os.walk(IMAGE_FILE_PATH):
+        current_subdir = root
+        new_subdir = []
+        new_subdir_illum = []
+        for filename in os.listdir(current_subdir):
+            match = re.search(filename_reg, filename)
+            if match:
+               new_subdir.append(os.path.join(root, match.group(1) + match.group(2) + match.group(3) + match.group(4)))
```

Comment:
The indentation here and below is funky. Always indent by four spaces! And limit text to 80c width. I think PyCharm (and most Python editors) have settings to warn you about this sort of stuff automatically.